Tests/GRPCHTTP2CoreTests/GRPCStreamStateMachineTests.swift

```diff
@@ -2686,7 +2659,145 @@ final class GRPCStreamServerStateMachineTests: XCTestCase {
+    var state = self.makeServerStateMachine(targetState: .clientIdleServerIdle)
+    let action1 = try state.receive(headers: .clientInitialMetadata, endStream: false)
+
+    let metadata: Metadata = [
+      ":path": "/test/test",
+      ":scheme": "http",
+      ":method": "POST",
+      "content-type": "application/grpc",
+      "te": "trailers",
+    ]
+    let descriptor = MethodDescriptor(service: "test", method: "test")
+    XCTAssertEqual(action1, .receivedMetadata(metadata, descriptor))
+
+    let action2 = try state.send(status: Status(code: .ok, message: ""), metadata: [:])
+    XCTAssertEqual(action2, .serverTrailers)
+
+    for endStream in [false, true] {
+      let action = try state.receive(buffer: ByteBuffer(), endStream: endStream)
+      XCTAssertEqual(action, .doNothing)
+    }
+  }
+
+  func testServerCantReadInboundAfterClosing() throws {
+    var state = self.makeServerStateMachine(targetState: .clientIdleServerIdle)
+    let action1 = try state.receive(headers: .clientInitialMetadata, endStream: false)
+
+    let metadata: Metadata = [
+      ":path": "/test/test",
+      ":scheme": "http",
+      ":method": "POST",
+      "content-type": "application/grpc",
+      "te": "trailers",
+    ]
+    let descriptor = MethodDescriptor(service: "test", method: "test")
+    XCTAssertEqual(action1, .receivedMetadata(metadata, descriptor))
+
+    // 10 empty message, each message is empty but needs a 5-byte header indicating that.
+    let buffer = ByteBuffer(repeating: 0, count: 5 * 10)
+
+    // Receive the buffer and read a message.
+    XCTAssertEqual(try state.receive(buffer: buffer, endStream: false), .readInbound)
+    XCTAssertEqual(state.nextInboundMessage(), .receiveMessage([]))
+
+    // Close the stream early.
+    let action2 = try state.send(status: Status(code: .ok, message: ""), metadata: [:])
+    XCTAssertEqual(action2, .serverTrailers)
+
+    // Try reading inbound.
+    XCTAssertEqual(state.nextInboundMessage(), .doNothing)
+  }
+
+  func testServerReceivesMultipleClientEncodings() throws {
+    var state = self.makeServerStateMachine(targetState: .clientIdleServerIdle)
+    let headers: HPACKHeaders = [
+      ":path": "/test/test",
+      ":scheme": "http",
+      ":method": "POST",
+      "content-type": "application/grpc",
+      "te": "trailers",
+      "grpc-encoding": "gzip,deflate",
+    ]
+
+    let action1 = try state.receive(headers: headers, endStream: false)
+    let trailers: HPACKHeaders = [
+      ":status": "200",
+      "content-type": "application/grpc",
+      "grpc-status": "13",
+      "grpc-message": "encoding must contain no more than one value.",
+    ]
+
+    XCTAssertEqual(action1, .rejectRPC(trailers: trailers))
+
+    // Extra input should be ignored.
+    for endStream in [false, true] {
+      let action = try state.receive(buffer: ByteBuffer(), endStream: endStream)
+      XCTAssertEqual(action, .doNothing)
+    }
+  }
+
+  func testServerReceivesUnsupportedCompression() throws {
```

Comment:
I think we already have this test: `testReceiveMetadataWhenClientIdleAndServerIdle_ServerUnsupportedEncoding`.
We're not asserting the last bit about extra input being ignored though.

Reply:
Ah cool, I'll remove this test.